### PR TITLE
Add LazyConfigAcceptor API

### PR DESCRIPTION
**Pull request by djc**
Based on https://github.com/ctz/rustls/pull/787. @jeromegn this compiles, want to give it a shot?

### CODE REVIEW

**Comment by jeromegn** [comment]
@djc thanks! That looks like it's going to work. Of course there's a lot of refactoring I need to do before I can really test it, but I wanted to quickly mention that it seemed :+1:

File: tokio-rustls/src/lib.rs [hunk 1]
```diff
@@ -126,6 +146,122 @@ impl TlsAcceptor {
...
+
+impl<IO> Accepted<IO>
+where
+    IO: AsyncRead + AsyncWrite + Unpin,
+{
+    pub fn client_hello(&self) -> rustls::ClientHello<'_> {
+        self.accepted.client_hello()
+    }
+
+    pub fn into_stream<F>(self, config: Arc<ServerConfig>) -> Accept<IO> {
```
Comment (jeromegn) [inline tokio-rustls/src/lib.rs:236]: Is there a way around `<F>` here? Rust cannot infer the type for `F` when using `into_stream`.

Comment (djc) [reply tokio-rustls/src/lib.rs:236]: Oops, I should remove that. You can use the other method for now with a trivial closure (see the body of this one).

File: tokio-rustls/src/lib.rs [hunk 2]
```diff
@@ -126,6 +146,122 @@ impl TlsAcceptor {
...
+                    return Poll::Ready(Err(io::Error::new(io::ErrorKind::InvalidInput, err)))
+                }
+            }
+        }
+
+        Poll::Pending
+    }
+}
+
+pub struct Accepted<IO> {
```
Comment (Matthias247) [inline tokio-rustls/src/lib.rs:223]: Naming nit: Is this really `Accepted`? Accepted is for me the state where we accepted a TCP connection. `Connected` is when we finished the handshake. Maybe `Handshaking/Handshaking/HandshakeStart/HandshakeRequest`?

File: tokio-rustls/src/lib.rs [hunk 3]
```diff
@@ -126,6 +146,122 @@ impl TlsAcceptor {
...
+            Err(error) => {
+                return Accept(MidHandshake::Error {
+                    io: self.io,
+                    // TODO(eliza): should this really return an `io::Error`?
+                    // Probably not...
+                    error: io::Error::new(io::ErrorKind::Other, error),
+                });
+            }
+        };
+        f(&mut conn);
```
Comment (Matthias247) [inline tokio-rustls/src/lib.rs:255]: Whats the use-case for the function? All the user to set further properties on the connection?

File: tokio-rustls/src/lib.rs (hunk 1 above)
Comment (djc) [reply tokio-rustls/src/lib.rs]: Removed it.

File: tokio-rustls/src/lib.rs (hunk 3 above)
Comment (djc) [reply tokio-rustls/src/lib.rs:255]: No clue, I just copied what `TlsAcceptor::accept_with()` does.

File: tokio-rustls/src/lib.rs (hunk 2 above)
Comment (djc) [reply tokio-rustls/src/lib.rs]: Renamed it to `StartHandshake` for now.

**Comment by jeromegn** [comment]
I'm not sure this is working right? I tried out a branch of our proxy that uses this and it appears handshakes are sometimes just plain stuck. They work other times (rarely).

I see some tests are failing. Maybe that's an indication 😄. All `tests/badssl.rs` tests are failing for me locally.

**Comment by jeromegn** [comment]
It appears the TLSv1.2 in the badssl tests sometimes work.

I've sprinkled some `println!`.

Bad run:

```
$ cargo test --package tokio-rustls --test badssl --all-features -- test_tls12 --exact --nocapture
    Finished test [unoptimized + debuginfo] target(s) in 0.02s
     Running tests/badssl.rs (target/debug/deps/badssl-047ddaa2d2987bee)

running 1 test
would block, prev: 32, buf.remaining(): 32, would_block: true, self.eof: false
would block, prev: 32, buf.remaining(): 32, would_block: true, self.eof: false
wants_read ok: 829 bytes
read: got 32 bytes
read: breaking! eof: true, would_block: false
tlsv12 output: HTTP/1.1 200 OK
Server: nginx/1
thread 'test_tls12' panicked at 'assertion failed: output.contains(\"<title>tls-v1-2.badssl.com</title>\")', tokio-rustls/tests/badssl.rs:57:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test test_tls12 ... FAILED
```

Good run:

```
$ cargo test --package tokio-rustls --test badssl --all-features -- test_tls12 --exact --nocapture
    Finished test [unoptimized + debuginfo] target(s) in 0.03s
     Running tests/badssl.rs (target/debug/deps/badssl-047ddaa2d2987bee)

running 1 test
would block, prev: 32, buf.remaining(): 32, would_block: true, self.eof: false
would block, prev: 32, buf.remaining(): 32, would_block: true, self.eof: false
wants_read ok: 798 bytes
read: got 32 bytes
read: got 32 bytes
read: got 64 bytes
read: got 128 bytes
read: got 256 bytes
read: got 257 bytes
wants_read ok: 31 bytes
read: got 0 bytes
read: breaking! eof: true, would_block: false
tlsv12 output: HTTP/1.1 200 OK
Server: nginx/1.10.3 (Ubuntu)
Date: Thu, 29 Jul 2021 17:40:48 GMT
Content-Type: text/html
Content-Length: 502
Last-Modified: Tue, 23 Feb 2021 21:28:41 GMT
Connection: close
ETag: "60357389-1f6"
Cache-Control: no-store
Accept-Ranges: bytes

<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <link rel="shortcut icon" href="/icons/favicon-green.ico"/>
  <link rel="apple-touch-icon" href="/icons/icon-green.png"/>
  <title>tls-v1-2.badssl.com</title>
  <link rel="stylesheet" href="/style.css">
  <style>body { background: green; }</style>
</head>
<body>
<div id="content">
  <h1 style="font-size: 12vw;">
    tls-v1-2.<br>badssl.com
  </h1>
</div>

</body>
</html>

test test_tls12 ... ok
```

It looks like it might give up reading too early in most cases?

I realize my earlier problems are with a TLS server, not with the client. I hope this issue might be related.

**Comment by djc** [comment]
Can you try your server again? There was a bug in there where I failed to reset the amount of buffer used if the buffer was fully used, so any `ClientHello` longer than 512 bytes would probably have hanged. If it still doesn't work, would also be useful to try with the branch from #68 to check whether the issue is specific to the `Acceptor` API or is due to other changes in rustls 0.20.

As for the badssl test failures, I'm unable to reproduce them locally. It looks like they might also be fixed on CI now. Can you see if they still reproduce for you?

**Comment by jeromegn** [comment]
This is still hanging for me. I'm testing on a small server doing about 50-60 TLS handshakes per second, from various kinds of clients.

Testing with cURL, it hangs here:
```
$ curl -iv https://<redacted>
* Rebuilt URL to: https://<redacted>/
*   Trying <redacted ipv6>...
* TCP_NODELAY set
* Connected to <redacted> (<redacted ipv6>) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
```

My code uses a `tower` service for handling I/O that needs a TLS handshake and it roughly looks like this:
- `LazyConfigAcceptor::new(rustls::Acceptor::new()?, io)`
- find the correct server config to use (simple logic)
- fetch the certificate we need (if not present) and add it to our cert resolver (this code did not change with the refactor)
- `accepted.into_stream(server_config)`

As for the tests, they still almost-always fail locally. Since this is hitting a remote resource, I wonder if it's latency-sensitive?

Failure:

```
$ cargo test --package tokio-rustls --test badssl --all-features -- test_tls12 --exact --nocapture
    Finished test [unoptimized + debuginfo] target(s) in 0.02s
     Running tests/badssl.rs (target/debug/deps/badssl-047ddaa2d2987bee)

running 1 test
would block, prev: 32, buf.remaining(): 32, would_block: true, self.eof: false
would block, prev: 32, buf.remaining(): 32, would_block: true, self.eof: false
wants_read ok: 829 bytes
read: got 32 bytes
read: breaking! eof: true, would_block: false
tlsv12 output: HTTP/1.1 200 OK
Server: nginx/1
thread 'test_tls12' panicked at 'assertion failed: output.contains(\"<title>tls-v1-2.badssl.com</title>\")', tokio-rustls/tests/badssl.rs:57:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test test_tls12 ... FAILED
```

Success:

```
$ cargo test --package tokio-rustls --test badssl --all-features -- test_tls12 --exact --nocapture
    Finished test [unoptimized + debuginfo] target(s) in 0.02s
     Running tests/badssl.rs (target/debug/deps/badssl-047ddaa2d2987bee)

running 1 test
would block, prev: 32, buf.remaining(): 32, would_block: true, self.eof: false
would block, prev: 32, buf.remaining(): 32, would_block: true, self.eof: false
wants_read ok: 798 bytes
read: got 32 bytes
read: got 32 bytes
read: got 64 bytes
read: got 128 bytes
read: got 256 bytes
read: got 257 bytes
would block, prev: 512, buf.remaining(): 255, would_block: true, self.eof: false
would block, prev: 255, buf.remaining(): 255, would_block: true, self.eof: false
would block, prev: 255, buf.remaining(): 255, would_block: true, self.eof: false
wants_read ok: 31 bytes
read: got 0 bytes
read: breaking! eof: true, would_block: false
tlsv12 output: HTTP/1.1 200 OK
Server: nginx/1.10.3 (Ubuntu)
Date: Fri, 30 Jul 2021 12:33:49 GMT
Content-Type: text/html
Content-Length: 502
Last-Modified: Tue, 23 Feb 2021 21:28:41 GMT
Connection: close
ETag: "60357389-1f6"
Cache-Control: no-store
Accept-Ranges: bytes

<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <link rel="shortcut icon" href="/icons/favicon-green.ico"/>
  <link rel="apple-touch-icon" href="/icons/icon-green.png"/>
  <title>tls-v1-2.badssl.com</title>
  <link rel="stylesheet" href="/style.css">
  <style>body { background: green; }</style>
</head>
<body>
<div id="content">
  <h1 style="font-size: 12vw;">
    tls-v1-2.<br>badssl.com
  </h1>
</div>

</body>
</html>

test test_tls12 ... ok
```

**Comment by jeromegn** [comment]
I can't reproduce my server issues locally if I replicate as close as I can our staging setup. So this is probably some deadlock happening somewhere due to conditions I can't reproduce in simple tests.

**Comment by djc** [comment]
I did still find some issues in how I buffered the data. I've now written a test to help flush out these issues, which passes. Wanna try again?

**Comment by jeromegn** [comment]
Thanks! Just tried again. It works for a little while, but then I see a large amount of deadline exceeded errors (we timeout the `Accepted` future after 5 seconds and the `Accept` future after 20 seconds). 

We have various "limits" in place so we don't overload our shared proxies. These work with semaphores, locks and queues! I think the issue might be there due to a subtle change in my refactor. (that's my current theory because we have limits per SNI and I noticed while a handshake with a specific SNI may be stuck entirely, another one still works fine)

I'm on the tokio discord is you'd rather chat in real-time.

**Comment by jeromegn** [comment]
I was able to reproduce my issue by running testssl.sh on this adapted example server:

```rust
use argh::FromArgs;
use rustls_pemfile::{certs, rsa_private_keys};
use std::fs::File;
use std::io::{self, BufReader};
use std::net::ToSocketAddrs;
use std::path::{Path, PathBuf};
use std::sync::Arc;
use tokio::io::{copy, sink, split, AsyncWriteExt};
use tokio::net::TcpListener;
use tokio_rustls::rustls::{self, Certificate, PrivateKey};
use tokio_rustls::LazyConfigAcceptor;

/// Tokio Rustls server example
#[derive(FromArgs)]
struct Options {
    /// bind addr
    #[argh(positional)]
    addr: String,

    /// cert file
    #[argh(option, short = 'c')]
    cert: PathBuf,

    /// key file
    #[argh(option, short = 'k')]
    key: PathBuf,

    /// echo mode
    #[argh(switch, short = 'e')]
    echo_mode: bool,
}

fn load_certs(path: &Path) -> io::Result<Vec<Certificate>> {
    certs(&mut BufReader::new(File::open(path)?))
        .map_err(|_| io::Error::new(io::ErrorKind::InvalidInput, "invalid cert"))
        .map(|mut certs| certs.drain(..).map(Certificate).collect())
}

fn load_keys(path: &Path) -> io::Result<Vec<PrivateKey>> {
    rsa_private_keys(&mut BufReader::new(File::open(path)?))
        .map_err(|_| io::Error::new(io::ErrorKind::InvalidInput, "invalid key"))
        .map(|mut keys| keys.drain(..).map(PrivateKey).collect())
}

#[tokio::main]
async fn main() -> io::Result<()> {
    let options: Options = argh::from_env();

    let addr = options
        .addr
        .to_socket_addrs()?
        .next()
        .ok_or_else(|| io::Error::from(io::ErrorKind::AddrNotAvailable))?;
    let certs = load_certs(&options.cert)?;
    let mut keys = load_keys(&options.key)?;
    let flag_echo = options.echo_mode;

    let config = Arc::new(
        rustls::ServerConfig::builder()
            .with_safe_defaults()
            .with_no_client_auth()
            .with_single_cert(certs, keys.remove(0))
            .map_err(|err| io::Error::new(io::ErrorKind::InvalidInput, err))?,
    );

    let listener = TcpListener::bind(&addr).await?;

    loop {
        let (stream, peer_addr) = listener.accept().await?;
        let config = config.clone();

        let fut = async move {
            let accepted =
                LazyConfigAcceptor::new(rustls::Acceptor::new().unwrap(), stream).await?;

            let mut stream = accepted.into_stream(config).await?;

            if flag_echo {
                let (mut reader, mut writer) = split(stream);
                let n = copy(&mut reader, &mut writer).await?;
                writer.flush().await?;
                println!("Echo: {} - {}", peer_addr, n);
            } else {
                let mut output = sink();
                stream
                    .write_all(
                        &b"HTTP/1.0 200 ok\r\n\
                    Connection: close\r\n\
                    Content-length: 12\r\n\
                    \r\n\
                    Hello world!"[..],
                    )
                    .await?;
                stream.shutdown().await?;
                copy(&mut stream, &mut output).await?;
                println!("Hello: {}", peer_addr);
            }

            Ok(()) as io::Result<()>
        };

        tokio::spawn(async move {
            if let Err(err) = fut.await {
                eprintln!("{:?}", err);
            }
        });
    }
}
```

The same appears to be true for the (untouched) `server` example.

**Review by quininer — APPROVED** [review]
LGTM

**Comment by djc** [comment]
@quininer can you release 0.23.1 with this?

**Comment by quininer** [comment]
@djc  Yes, I will release a new version later.

**Pull request merged**